draft-ietf-quic-tls.md

```diff
@@ -724,6 +689,37 @@ This results in abandoning loss recovery state for the Initial encryption level
 and ignoring any outstanding Initial packets.
 
 
+### Discarding Handshake Keys
+
+In this document, the TLS handshake is considered confirmed from an endpoint's
+perspective when both of the following two conditions are met: the handshake is
+complete from this endpoint’s perspective and the endpoint has received an
+ACK for a packet sent with 1-RTT keys.  This second condition can be
+implemented as checking whether largest_recvd_ack_1rtt >= first_1rtt_pn, where
+largest_recvd_ack_1rtt represents the highest value of the Largest Acknowledged
+field in any received 1-RTT ACK frame, and where first_1rtt_pn represents the
+lowest packet number that this endpoint has sent encrypted with 1-RTT keys
+(first_1rtt_pn is always 0 on the server since servers cannot send 0-RTT
+packets).
+
+An endpoint MUST NOT discard its handshake keys until the TLS handshake is
+confirmed from its perspective.  An endpoint MAY discard its handshake keys
+as soon as it has confirmed the handshake.  Endpoints SHOULD repeatedly send
+ACK-eliciting frames encrypted with 1-RTT keys as soon as they have installed
+those send keys and until they receive an acknowledgment for one of them.
+
+
+### Discarding 0-RTT Keys
+
+Clients can discard 0-RTT keys as soon as they install 1-RTT keys, since they
+have no use after that moment.
+
+Servers SHOULD discard 0-RTT keys as soon as they no longer expect to receive
+0-RTT packets.  This can been established by receiving a 1-RTT packet when all
+previous packet numbers have been received, or by receiving a 1-RTT packet and
+then waiting until three times the Probe Timeout (PTO, see {{QUIC-RECOVERY}}).
```

Comment:
We should talk about this.  If knowing the 0-RTT status of packets is important, then it might make sense to drop 0-RTT packets that are reordered to prevent confusing them with 1-RTT.  In our DTLS 0-RTT implementation we've chosen to drop 0-RTT packets as soon as we start accepting 1-RTT packets.  That's something of an implementation choice, but managing the alternative is really tricky because every read needs to be annotated carefully and applications need to ask about every bit.

Reply:
That makes sense to me, but I'm personally OK with giving implementors leniency since this is a SHOULD. Do you have suggestions on how to phrase this better?

Reply:
TLS 1.3 is fairly straightforward since it's all ordered. We toss the 0-RTT keys as soon as we see/send EndOfEarlyData. That probably wouldn't apply for QUIC.

(I haven't paid a whole lot of attention to this issue and don't have any immediate opinion. I'm happy to chat with folks about it if it'd be useful, and I'm also happy just leaving it in your capable hands and not add too many cooks.)

Reply:
Yeah, TLS is simple.  I thought that you might have thought about what it might look like for DTLS, which is much more like QUIC.

Reply:
Nah, I haven't really done anything with DTLS. I [read through](https://mailarchive.ietf.org/arch/msg/tls/VNrSd7gv7uFjJNZH6frBt5lb57A) the spec the other day, but that's about it.

Reply:
I've modified the text to MUST discard, and allow waiting up to 3 PTO